lib/grizzly/virtual_devices_registry.ex

```diff
@@ -0,0 +1,95 @@
+defmodule Grizzly.VirtualDevicesRegistry do
+  @moduledoc false
+
+  alias Grizzly.VirtualDevices
+  alias Grizzly.VirtualDevices.Device
+  alias Grizzly.ZWave.DeviceClass
+
+  @typedoc """
+  """
+  @type device_entry() :: %{
+          device_impl: Device.t(),
+          device_class: DeviceClass.t(),
+          device_opts: [Device.device_opt()],
+          id: VirtualDevices.id(),
+          pid: pid()
+        }
+
+  @type start_opts() :: [
+          {:keys, :unique},
+          {:name, module()},
+          meta: [{:inclusion_handler, Grizzly.handler()}]
+        ]
+
+  @doc """
+  Function to return the options for starting this registry
+  """
+  @spec start_options(Grizzly.Options.t()) :: start_opts()
+  def start_options(grizzly_opts) do
+    [keys: :unique, name: __MODULE__, meta: [inclusion_handler: grizzly_opts.inclusion_handler]]
+  end
+
+  @doc """
+  Register the virtual device
+  """
+  @spec register(Device.t(), DeviceClass.t(), [Device.device_opt()]) :: device_entry()
+  def register(device_impl, device_class, device_opts) do
+    id = {:virtual, Registry.count(__MODULE__) + 1}
+
+    entry = %{
+      device_impl: device_impl,
+      device_class: device_class,
+      id: id,
+      device_opts: device_opts
+    }
+
+    case Registry.register(__MODULE__, id, entry) do
+      {:ok, pid} ->
+        Map.put(entry, :pid, pid)
+
+      {:error, {:already_registered, pid}} ->
+        Map.put(entry, :pid, pid)
+    end
+  end
+
+  @doc """
+  Unregister a device from the registry
+  """
+  @spec unregister(VirtualDevices.id()) :: :ok
+  def unregister(device_id) do
+    Registry.unregister(__MODULE__, device_id)
+  end
+
+  @doc """
+  Get a device entry by the virtual device id
+  """
+  @spec get(VirtualDevices.id()) :: device_entry() | nil
+  def get(device_id) do
+    case Registry.lookup(__MODULE__, device_id) do
+      [] ->
+        nil
+
+      [{_pid, entry}] ->
+        Map.put(entry, :id, device_id)
```

Comment:
I don't understand why the Map.put is necessary since the id is set at the moment of registration.